### PR TITLE
feat: add EAGLE-1 / EAGLE-2 drafter training backend

### PR DESCRIPTION
**Pull request by khazic**
## What

Adds an EAGLE-1 / EAGLE-2 drafter training backend to the SpeCo overlay, closing [#14](https://github.com/verl-project/verl-SpeCo/issues/14). The `Draft Model Support` matrix now covers EAGLE-1, EAGLE-2, EAGLE-3, DFLASH and DSpark.

Select it with:

```
actor_rollout_ref.rollout.drafter.speculative_algorithm=EAGLE1   # or EAGLE2
```

## Design

The training objective follows the EAGLE-1/2 formulation (single-step feature prediction), which is unchanged between the two algorithms; EAGLE-2 differs from EAGLE-1 only in the inference-time speculative tree policy (a dynamic, context-aware tree), which lives in the rollout engine, not in drafter training. Both map to vLLM's native `eagle` method.

The dense draft (`verl_speco/models/eagle1/`) fuses the token embedding with the target's last-layer hidden state through a single `fc` layer (`hidden_size + target_hidden_size -> hidden_size`), runs one standard decoder layer, and predicts the target's next-step hidden state. It carries no `lm_head`: token logits come from the frozen target head (weight tying), matching the original EAGLE design. The loss combines a SmoothL1 feature-regression term (weight 1.0) with a full-vocabulary soft cross-entropy distillation term (weight 0.1), plus optional EAGLE feature-noise augmentation.

`Eagle1TrainerBackend` reuses the EAGLE-3 online data-collection plumbing by reporting `model_type == "eagle3"`, so `base_trainer` assembles the shifted inputs and the `last_hidden_states` regression target; only `build_model` and `compute_loss` differ. Old-logprob hidden collection captures a single final target layer for EAGLE-1/2 (versus EAGLE-3's low/mid/high triple). The soft cross-entropy reuses the EAGLE-3 helper, so non-finite logits are sanitized before `log_softmax` and masked positions cannot backprop NaN gradients.

## Not a duplicate

Checked `gh issue view 14` (open, unassigned, no competing work) and the open PRs (`gh pr list`). #13 (Separate Draft Model Training) and #10 (draft weight loading) are unrelated; no open PR touches EAGLE-1/2.

## Testing

A human reviewed every changed line and ran the tests below.

CPU contract tests (`tests/integration/test_eagle1_backend_contract.py`), 12 passed: draft forward shape, loss equivalence against the reference EAGLE-1 formula, the `use_logits=False` requirement, config-time rejection of `use_logits=True`, the Ulysses-SP opt-out, EAGLE-1/EAGLE-2 to `eagle` vLLM method mapping, and single-final-layer aux selection (including ignoring a stray multi-layer config). The full suite is `100 passed, 2 skipped, 2 failed`; both failures pre-exist on `main` (a `tests/heavy` runner-env check and a stale metrics assertion) and are unrelated to this change.

```
pytest tests/integration/test_eagle1_backend_contract.py -q
# 12 passed
```

GPU hardware smoke (`ci/eagle1_gpu_smoke.py`) on a real target (Qwen3-4B, 1x A100): it runs the frozen target forward to produce genuine last-layer hidden states, assembles the batch exactly as `base_trainer` does for the EAGLE shifted path, builds the draft via `Eagle1TrainerBackend.build_model`, and runs 100 optimizer steps. The draft is cold-started, so the signal is convergence: SmoothL1 feature loss and soft-CE token loss fall and draft-vs-target top-1 agreement rises. EAGLE-2 produces identical numbers (same training path).

```
python ci/eagle1_gpu_smoke.py --target /path/to/Qwen3-4B --algorithm EAGLE1 --steps 100
```

<details>
<summary>EAGLE-1 GPU training log (Qwen3-4B target, cold start, 100 steps)</summary>

```
[smoke] loading target /path/to/Qwen3-4B
[smoke] batch seq_len=66 hidden=2560
[smoke] algorithm=EAGLE1 draft_layers=1 fc=5120->2560 trainable_params=114,040,320
[smoke] step   0  hidden_loss=1.4905  token_loss=12.1883  loss=2.7094  draft_vs_target_top1=0.1364
[smoke] step  10  hidden_loss=1.1124  token_loss=2.7478  loss=1.3872  draft_vs_target_top1=0.8485
[smoke] step  20  hidden_loss=0.9030  token_loss=2.2944  loss=1.1324  draft_vs_target_top1=0.8636
[smoke] step  30  hidden_loss=0.7947  token_loss=2.1330  loss=1.0080  draft_vs_target_top1=0.9394
[smoke] step  40  hidden_loss=0.7217  token_loss=2.0425  loss=0.9260  draft_vs_target_top1=0.9545
[smoke] step  50  hidden_loss=0.6683  token_loss=2.0004  loss=0.8683  draft_vs_target_top1=0.9394
[smoke] step  60  hidden_loss=0.6311  token_loss=1.9821  loss=0.8293  draft_vs_target_top1=0.9545
[smoke] step  70  hidden_loss=0.6082  token_loss=1.9735  loss=0.8055  draft_vs_target_top1=0.9545
[smoke] step  80  hidden_loss=0.5965  token_loss=1.9690  loss=0.7934  draft_vs_target_top1=0.9545
[smoke] step  90  hidden_loss=0.5918  token_loss=1.9644  loss=0.7883  draft_vs_target_top1=0.9545
[smoke] step  99  hidden_loss=0.5903  token_loss=1.9612  loss=0.7864  draft_vs_target_top1=0.9545
[smoke] DONE  hidden_loss 1.4905->0.5903  token_loss 12.1883->1.9612  top1=0.9545
```

</details>

<details>
<summary>EAGLE-2 GPU training log (Qwen3-4B target, cold start, 100 steps) — identical training path</summary>

```
[smoke] loading target /path/to/Qwen3-4B
[smoke] batch seq_len=66 hidden=2560
[smoke] algorithm=EAGLE2 draft_layers=1 fc=5120->2560 trainable_params=114,040,320
[smoke] step   0  hidden_loss=1.4905  token_loss=12.1883  loss=2.7094  draft_vs_target_top1=0.1364
[smoke] step  10  hidden_loss=1.1124  token_loss=2.7478  loss=1.3872  draft_vs_target_top1=0.8485
[smoke] step  20  hidden_loss=0.9030  token_loss=2.2944  loss=1.1324  draft_vs_target_top1=0.8636
[smoke] step  30  hidden_loss=0.7947  token_loss=2.1330  loss=1.0080  draft_vs_target_top1=0.9394
[smoke] step  40  hidden_loss=0.7217  token_loss=2.0425  loss=0.9260  draft_vs_target_top1=0.9545
[smoke] step  50  hidden_loss=0.6683  token_loss=2.0004  loss=0.8683  draft_vs_target_top1=0.9394
[smoke] step  60  hidden_loss=0.6311  token_loss=1.9821  loss=0.8293  draft_vs_target_top1=0.9545
[smoke] step  70  hidden_loss=0.6082  token_loss=1.9735  loss=0.8055  draft_vs_target_top1=0.9545
[smoke] step  80  hidden_loss=0.5965  token_loss=1.9690  loss=0.7934  draft_vs_target_top1=0.9545
[smoke] step  90  hidden_loss=0.5918  token_loss=1.9644  loss=0.7883  draft_vs_target_top1=0.9545
[smoke] step  99  hidden_loss=0.5903  token_loss=1.9612  loss=0.7864  draft_vs_target_top1=0.9545
[smoke] DONE  hidden_loss 1.4905->0.5903  token_loss 12.1883->1.9612  top1=0.9545
```

</details>

## Known follow-ups

- Ulysses sequence parallelism is not yet implemented for the EAGLE-1/2 loss; the backend opts out so it is forced off rather than aborting.
- The draft uses eager attention (as in the reference implementation); a flex-attention path like EAGLE-3's could be added for long sequences.
- SGLang rollout wiring for EAGLE-1/2 can follow; this PR targets the vLLM path.

### CODE REVIEW

File: verl_speco/backends/eagle1_trainer_backend.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+            attention_mask=attention_mask,
+            position_ids=position_ids,
+        )
+
+        # Weight-tied logits from the frozen target head. Gradients flow into the
+        # draft through the frozen linear; the head itself is not updated.
+        predicted_logits = self.target_model(predicted_hidden).float()
+        with torch.no_grad():
+            target_logits = self.target_model(last_hidden_states).float()
+            target_probs = torch.softmax(target_logits, dim=-1)
```
[the remark this answers is not in the document]
Comment (khazic) [reply verl_speco/backends/eagle1_trainer_backend.py:165]: The eagle3 backend it mirrors calls `self.target_model(...)` directly (eagle3_trainer_backend.py:964,988), and both possible heads (_SyncedTargetHead and TargetHead) are nn.Modules with a forward. The GPU smoke exercises the real TargetHead path this way and reaches top1=0.95, so a direct call is correct. Keeping it consistent with eagle3.

File: verl_speco/backends/eagle1_trainer_backend.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        draft_config = self._build_draft_config(spec_model_path, target_hf_config)
+        if not hasattr(draft_config, "target_hidden_size"):
+            draft_config.target_hidden_size = int(target_hf_config.hidden_size)
+        self.vocab_size = draft_config.vocab_size
+
+        if spec_model_path and os.path.exists(os.path.join(spec_model_path, "config.json")):
+            log_drafter_checkpoint_step(logger, spec_model_path, action="Loading EAGLE-1/2 drafter weights")
+            drafter_module = LlamaForCausalLMEagle1.from_pretrained(spec_model_path, config=draft_config)
+        else:
+            drafter_module = LlamaForCausalLMEagle1(draft_config)
```
[the remark this answers is not in the document]
Comment (khazic) [reply verl_speco/backends/eagle1_trainer_backend.py:107]: Same behavior as the eagle3 backend, which also gates weight loading on a local path (eagle3_trainer_backend.py:644). Cold start is the common case here, and hub-id draft init is a separate enhancement that would apply to both backends; I'd rather not add the ad-hoc 'path/to' heuristic. Left as a follow-up.

File: ci/eagle1_gpu_smoke.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        if step % 10 == 0 or step == args.steps - 1:
+            with torch.no_grad():
+                predicted_hidden = draft(
+                    input_ids=batch["input_ids"],
+                    hidden_states=batch["hidden_states"],
+                    attention_mask=batch["attention_mask"],
+                    position_ids=batch["position_ids"],
+                )
+                pred_logits = backend.target_model(predicted_hidden).float()
+                tgt_logits = backend.target_model(batch["last_hidden_states"]).float()
```
[the remark this answers is not in the document]
Comment (khazic) [reply ci/eagle1_gpu_smoke.py:147]: backend.target_model is the real head built by build_model (TargetHead / _SyncedTargetHead), which is callable; the smoke already runs this path successfully. Same rationale as the backend comment.

File: verl_speco/backends/eagle1_trainer_backend.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                draft_top1 = predicted_logits.argmax(dim=-1)
+                target_top1 = target_probs.argmax(dim=-1)
+                acc = ((draft_top1 == target_top1) & valid_mask).float().sum() / num_tokens.clamp_min(1)
+            logger.debug(
+                "[eagle1 loss] tokens=%s hidden_loss=%.6f token_loss=%.6f top1_acc=%.6f",
+                int(num_tokens.detach().cpu().item()),
+                float((total_local_vloss / num_tokens.clamp_min(1)).detach().cpu().item()),
+                float((total_local_ploss / num_tokens.clamp_min(1)).detach().cpu().item()),
+                float(acc.detach().cpu().item()),
+            )
```
[the remark this answers is not in the document]
Comment (khazic) [reply verl_speco/backends/eagle1_trainer_backend.py]: Good catch, done: the per-step block is now gated on logger.isEnabledFor(logging.DEBUG) so the .item() syncs are skipped at INFO+ (84f0867).

File: verl_speco/trainer/base_trainer.py [hunk 5]
```diff
@@ -471,6 +471,20 @@ def __init__(
...
+            # Backends that do not implement the SP loss path (e.g. EAGLE-1/2)
+            # opt out here so SP is never enabled under them, rather than aborting
+            # in compute_loss when rollout_tp > 1 on a multi-rank training group.
+            self.ulysses_sequence_parallel_size = 1
+            if rollout_tp_size > 1 and self.training_group_world_size > 1:
+                logger.debug(
+                    "[Rank %s] Disable Ulysses SP for %s drafter training: "
+                    "rollout_tp=%s training_group_world_size=%s",
+                    self.rank,
+                    self.backend.model_type,
```
[the remark this answers is not in the document]
Comment (khazic) [reply verl_speco/trainer/base_trainer.py]: Agreed, fixed: it now logs the actual speculative_algorithm (EAGLE1/EAGLE2) instead of the reused eagle3 model_type (84f0867).
